File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, node: Optional[TreeNode], lo: int, hi: int) -> int:
        if not node: return abs(hi - lo)
        lo = min(lo, node.val)
        hi = max(hi, node.val)
        return max(self.dfs(node.left, lo, hi), self.dfs(node.right, lo, hi))
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        return self.dfs(root, inf, -inf)

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int dfs(TreeNode* node, int min, int max) {
        if(node == nullptr) return abs(max - min);
        if(node->val < min) min = node->val;
        if(node->val > max) max = node->val;
        int a = dfs(node->left, min, max);
        int b = dfs(node->right, min, max);
        return a > b ? a : b;
    }
    int maxAncestorDiff(TreeNode* root) {
        return dfs(root, INT_MAX, INT_MIN);
    }
};